Mem 787 tutorial model run (#812)
CDCgov/PyRenew: docs/tutorials/building_multisignal_models.qmd

Reviewer: The comment isn't accurate. The print call just prints the length of the PMF. Seems worth fixing here since it's a quick fix. I thnk just remove any output, or just print the generation interval distribution itself b/c it's short.

Reviewer: ```suggestion
- **Initial infections ($(I)(0)$)**: Starting condition for the renewal process
```

Reviewer: ```suggestion
### Log $\mathcal{R}(t)$ at time $0$
```

Reviewer: ```suggestion
We place a prior on the $\log(\mathcal{R}(t))$ at time $t = 0$, centered at 0.0 ($\mathcal{R}(t=0) = 1$) with moderate uncertainty:
```

Reviewer: ```suggestion
### Choosing the \mathcal{R}(t)$ Parameter Cadence
```

Reviewer: ```suggestion
- **NumPyro sites**: Prefixes all sample and deterministic sites (e.g., `hospital_obs`, `hospital_predicted`)
```

Author: to render  .qmd to html you need a space before a bulleted list.   maybe not for .qmd to .md, but I see no harm in leaving it as is.

Reviewer: ```suggestion
The `Wastewater` observation process uses `LogNormalNoise`, which takes hierarchical priors for the site-level mode and standard deviation parameters.
```

Reviewer: ```suggestion
Here we specify `HierarchicalNormalPrior` for the site-level mode and `GammaGroupSdPrior` for the standard deviation.
```

Reviewer: Apologies for not noticing this sooner. I don't think we need (or want) the key to vary depending on when the tutorial is run, and wrapping this in a function in a collapsed codeblock also makes it harder for an inexperienced user to learn how to pass a key to `model.run()`. I think remove and replace with explicitly specified keys.

- **Model time axis**: the daily axis used by the renewal equation and delay convolutions
- **Observation cadence**: the temporal granularity for each signal, such as daily ED visits or weekly hospital admissions

> The DifferencedAR(1) process above samples one value per model day:

Reviewer: ```suggestion
The `DifferencedAR1` process above samples one value per model day:
```

Reviewer: either  `DifferencedAR1` (the class) or "AR(1) process on the first differences" the concept. I think the former is fine.

```

The `**obs_data` dictionary supplies the observation start date and one data dictionary per observation process.
The keys of the observation dictionaries are the names registered on the builder: `hospital` and `wastewater`.

### Model Time vs. Observation Time

In the raw data, day 0 is the first day with observations.
PyRenew adds an initialization period before this date so the renewal process has enough past infections to generate infections during the observed period.

This means the model has two related time scales:

- **Observation time** starts at the first day in the data.
- **Model time** starts `n_init` days earlier.

On the model time axis, indices `0` through `n_init - 1` are initialization days.
The first observed day is model index `n_init`, and the last observed day is model index `n_init + n_days - 1`.

The model provides helper methods to convert observation data onto this model time axis:

- `model.pad_observations(obs)` prepends `n_init` NaN values to dense daily observation vectors.
- `model.shift_times(times)` adds `n_init` to sparse observation time indices.

The hospital admissions signal is dense: it has one value per observation day, so we pad it.

> The wastewater signal is sparse: each row has an observation day, so we shift its time index.

Reviewer 2: On first read I assumed the model handles this automatically, and it took me a moment to realize the helper `prepare_observation_data` was doing padding/shifting. I think doing this data prep inline instead of a helper function is more illustrative

Author: The helper functions are part of the `MultiSignalModel` class - I could link these to the documentation - would that help?

The goal is to avoid making the user do their own time indexing because it's too easy to mess it up.

Reviewer 2: I see. Then is the reason for creating `prepare_observation_data()` as opposed to something like below to DRY data prep for two n_fit_days?
```
obs_data = {
        "obs_start_date": obs_start_date,
        "hospital": {
            "obs": model.pad_observations(hosp_admits[:n_days_fit]),
        },
        "wastewater": {
            "obs":  ww_data["observed_conc"][ww_mask],
            "times": model.shift_times(ww_data["time_indices"][ww_mask]),
            "subpop_indices": ww_subpop_indices,
            "sensor_indices": ww_data["site_ids"][ww_mask],
            "n_sensors": ww_data["n_sites"],
        },
    }

Reviewer 2: I am also curious about the reasoning to include/compare 90-day vs 180-day fits in the tutorial for building multi-signal model.

Author: good question.   when I  wrote this tutorial I was wondering why in production the models were run on 90 or 180 days worth of data, so I put this into the tutorial to see what would happen.
you're quite correct that this doesn't have anything to do with building the multisignal model, and it makes the tutorial rather slow to render. 
I could take out the 180 day fit and just leave the 90 day fit if you think its appropriate.

Reviewer 2: Yes, I like that idea! Just fit to 90 days to keep it more focused

Author: got rid of 180 day fit.  added more informative plot to illustrate uncertainty at the edge of the observation window.

Author: I also revised the data prep section and removed the now-unneeded `prepare_observation_data` and added more discussion of `pad_observations` and `shift_times`.

print(f"Total population: {float(jnp.sum(subpop_fractions)):.0%}")
```

The `subpop_fractions` array defines all subpopulations in the latent infection process, and its entries must sum to 1.0.
Which subpopulations each observation process sees is determined by the observation data.

> For wastewater, `subpop_indices` links each observation to the appropriate subpopulation, while `sensor_indices` selects the sensor-specific noise parameters.

Reviewer 2: Can we make this point about `subpop_indices` and `sensor_indices` more prominent, maybe bullets?

Author: improved discussion, bulleted `subpop_indices` and `sensor_indices`

    }
```

The returned dictionary has the following structure:

```python
{
    "obs_start_date": ...,
    "hospital": {
        "obs": ...
    },
    "wastewater": {
        "obs": ...
        "times": ...
        "subpop_indices": ...
        "sensor_indices": ...

>         "n_sensors": ...

Reviewer 2: The tutorial does not mention why `n_sensors` is required in the wastewater data dictionary

Author: added to discussion of the wastewater measurement observation paragraph starting on line 709.

}
```

> At call time, this dictionary is unpacked with `**`, forwarding each entry as a keyword argument to `model.run()`.

Reviewer: ```suggestion
At call time, we unpack this dictionary with `**`, forwarding each entry as a keyword argument to `model.run()`. Note that we could also have manually passed each dictionary entry to `model.run()`. We chose to create (and then unpack) an `obs_data` dictionary to keep our data organized in one place. 
```

)

> ```

Reviewer 2: What is the reason for showing 14 days Moving Average of Hospital Admissions instead of daily values?

Author: smoothing.  the daily values are plotted as points - the data is very noisy.  you want to see some kind of trend, and see that the trend lags the infections.

```{python}
#| label: monthly-CI

ci_width_90 = quantiles_90["q95"] - quantiles_90["q05"]

ci_width_df = pd.DataFrame(
    {
        "day": np.arange(n_days_90),
        "ci_width": ci_width_90,
        "period": pd.cut(
            np.arange(n_days_90),
            bins=[-1, 29, 59, 89],
            labels=["Days 0-29", "Days 30-59", "Days 60-89"],
        ),
    }
)

period_summary_df = (
    ci_width_df.groupby("period", observed=True)
    .agg(
        start=("day", "min"),
        end=("day", "max"),
        mean_ci_width=("ci_width", "mean"),
    )
    .reset_index()
)

print("\nMean 90% interval width by time period:")
for row in period_summary_df.itertuples(index=False):
    print(f"  {row.period}: {row.mean_ci_width:,.0f}")

(
    p9.ggplot(ci_width_df, p9.aes(x="day", y="ci_width"))
    + p9.geom_line(color="gray", alpha=0.6, size=0.7)
    + p9.geom_segment(
        data=period_summary_df,
        mapping=p9.aes(
            x="start",
            xend="end",
            y="mean_ci_width",
            yend="mean_ci_width",
            color="period",
        ),
        size=1.6,

>     )

Reviewer: This plot should also have a log scale y axis (consistent with the incidence plot above.

```suggestion
    )
   + p9.scale_y_log10()
```

    + p9.labs(
        x="Day",
        y="90% interval width",
        color="Period",

>         title="Posterior uncertainty over the 90-day fit",

Reviewer 2: ```suggestion
        title="Posterior uncertainty in latent infections over the 90-day fit",
```
